service/history/workflowcache/cache.go

```diff
@@ -0,0 +1,107 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package workflowcache
+
+import (
+	"time"
+
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/quotas"
+)
+
+// WFCache is a per workflow cache used for workflow specific in memory data
+type WFCache interface {
+	AllowExternal(domainID string, workflowID string) bool
+	AllowInternal(domainID string, workflowID string) bool
+}
+
+type wfCache struct {
+	lru                    cache.Cache
+	externalLimiterFactory quotas.LimiterFactory
+	internalLimiterFactory quotas.LimiterFactory
+}
+
+type cacheKey struct {
+	domainID   string
+	workflowID string
+}
+
+type cacheValue struct {
+	externalRateLimiter quotas.Limiter
+	internalRateLimiter quotas.Limiter
+}
+
+// Params is the parameters for a new WFCache
+type Params struct {
+	TTL                    time.Duration
+	MaxCount               int
+	ExternalLimiterFactory quotas.LimiterFactory
+	InternalLimiterFactory quotas.LimiterFactory
+}
+
+// New creates a new WFCache
+func New(params Params) WFCache {
+	return &wfCache{
+		lru: cache.New(&cache.Options{
+			TTL:      params.TTL,
+			Pin:      true,
+			MaxCount: params.MaxCount,
+		}),
+		externalLimiterFactory: params.ExternalLimiterFactory,
+		internalLimiterFactory: params.InternalLimiterFactory,
+	}
+}
+
+// AllowExternal returns true if the rate limiter for this domain/workflow allows an external request
+func (c *wfCache) AllowExternal(domainID string, workflowID string) bool {
+	// Locking is not needed because both getCacheItem and the rate limiter are thread safe
+	value := c.getCacheItem(domainID, workflowID)
+	return value.externalRateLimiter.Allow()
+}
+
+// AllowInternal returns true if the rate limiter for this domain/workflow allows an internal request
+func (c *wfCache) AllowInternal(domainID string, workflowID string) bool {
+	// Locking is not needed because both getCacheItem and the rate limiter are thread safe
+	value := c.getCacheItem(domainID, workflowID)
+	return value.internalRateLimiter.Allow()
+}
+
+func (c *wfCache) getCacheItem(domainID string, workflowID string) *cacheValue {
+	// The underlying lru cache is thread safe, so there is no need to lock
+	key := cacheKey{
+		domainID:   domainID,
+		workflowID: workflowID,
+	}
+
+	value, ok := c.lru.Get(key).(*cacheValue)
+
+	if !ok {
+		value = &cacheValue{
+			externalRateLimiter: c.externalLimiterFactory.GetLimiter(domainID),
+			internalRateLimiter: c.internalLimiterFactory.GetLimiter(domainID),
+		}
+		c.lru.PutIfNotExist(key, value)
```

Comment:
I think you need lock here because if 2 concurrent threads reach here only one of them will add the `value` to the cache and the other one will return its local `value` instead of the cached instance

Reply:
Resolved here https://github.com/uber/cadence/pull/5605